pallets/stablewap/src/lib.rs

```diff
@@ -0,0 +1,712 @@
+
+			T::Currency::transfer(assets.0, &who, &pool_account, initial_liquidity.0)?;
+			T::Currency::transfer(assets.1, &who, &pool_account, initial_liquidity.1)?;
+
+			T::Currency::deposit(pool_id.0, &who, share_amount)?;
+
+			Self::deposit_event(Event::PoolCreated {
+				id: pool_id,
+				assets,
+				initial_liquidity,
+				amplification,
+			});
+
+			Ok(())
+		}
+
+		/// Add liquidity to selected 2-asset pool.
+		///
+		/// LP must provide liquidity of both assets by specifying amount of asset a.
+		///
+		/// Amount of asset b is calculated so the ratio does not change:
+		///
+		/// new_reserve_b = (reserve_a + amount) * reserve_b / reserve_a
+		/// amount_b = new_reserve_b - reserve_b
+		///
+		/// LP must have sufficient amount of asset a/b - amount_a and amount_b.
+		///
+		/// Origin is given corresponding amount of shares.
+		///
+		/// Parameters:
+		/// - `origin`: liquidity provider
+		/// - `pool_id`: Pool Id
+		/// - `asset`: Asset id
+		/// - `amount`: liquidity amount of `asset` to be added to the pool.
+		///
+		/// Emits `LiquidityAdded` event when successful.
+		#[pallet::weight(<T as Config>::WeightInfo::add_liquidity())]
+		#[transactional]
+		pub fn add_liquidity(
+			origin: OriginFor<T>,
+			pool_id: PoolId<T::AssetId>,
+			asset: T::AssetId,
+			amount: Balance,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				amount >= T::MinTradingLimit::get(),
+				Error::<T>::InsufficientTradingAmount
+			);
+
+			// Ensure that who has enough balance of given asset
+			ensure!(
+				T::Currency::free_balance(asset, &who) >= amount,
+				Error::<T>::InsufficientBalance
+			);
+
+			let pool = Pools::<T>::get(pool_id).ok_or(Error::<T>::PoolNotFound)?;
+
+			ensure!(pool.contains_asset(asset), Error::<T>::AssetNotInPool);
+
+			// Load initial pool assets balances.
+			let pool_account = T::ShareAccountId::from_assets(&pool.assets, Some(POOL_IDENTIFIER));
+			let initial_reserves = AssetAmounts(
+				T::Currency::free_balance(pool.assets.0, &pool_account),
+				T::Currency::free_balance(pool.assets.1, &pool_account),
+			);
+
+			// Work out which asset is asset B ( second asset which has to provided by LP )
+			// Update initial reserves in correct order ( asset a is given as params, asset b is second asset)
+			let (assets, updated_a_reserve, initial_reserves) = if asset == pool.assets.0 {
+				let assets = (asset, pool.assets.1);
+				let reserves = AssetAmounts(initial_reserves.0, initial_reserves.1);
+				let updated_a_reserve = reserves.0.checked_add(amount).ok_or(ArithmeticError::Overflow)?;
+
+				(assets, updated_a_reserve, reserves)
+			} else {
+				let assets = (asset, pool.assets.0);
+				let reserves = AssetAmounts(initial_reserves.1, initial_reserves.0);
+				let updated_a_reserve = reserves.0.checked_add(amount).ok_or(ArithmeticError::Overflow)?;
+
+				(assets, updated_a_reserve, reserves)
+			};
```

Comment:
```suggestion
			// Work out which asset is asset B ( second asset which has to provided by LP )
			// Provide assets and initial reserves in correct order (asset a is given as params, asset b is second asset)
			let (assets, initial_reserves) = {
			  let initial = (
				  T::Currency::free_balance(pool.assets.0, &pool_account),
				  T::Currency::free_balance(pool.assets.1, &pool_account),
			  );
			  if asset == pool.assets.0 {
				  ((asset, pool.assets.1), AssetAmounts(initial.0, initial.1))
			  } else {
				  ((asset, pool.assets.0), AssetAmounts(initial.1, initial.0))
			  }
			};
			
			let updated_a_reserve = reserves.0.checked_add(amount).ok_or(ArithmeticError::Overflow)?;
```
What do you think of this?

Reply:
i like.